aurora_robot_tools/camera/process_image.py

```diff
@@ -0,0 +1,570 @@
+
+    def _get_transformation_matrix(self, centers: list[tuple]) -> np.array:
+        """ Takes center points of reference image and gets transformation matrix.
+
+        Args:
+            centers (list[tuple]): list with the four corner coordinates
+
+        Returns:
+            M (array): transformation matrix
+        """
+        pts2 = np.float32((self.mm_coords + self.offset_mm)*self.mm_to_pixel)
+        # Sort center coordinates in correct order for transformation matrix
+        y_values = [center[1] for center in centers] # Extract the y-values
+        mean_y = sum(y_values) / len(y_values) # Calculate the mean of the y-values
+        # Split the list based on the median y-value
+        lower_y_group = [center for center in centers if center[1] < (mean_y - 500)]
+        higher_y_group = [center for center in centers if center[1] > (mean_y + 500)]
+        # Sort top and bottom points by x
+        top_half_sorted = sorted(lower_y_group, key=lambda x: x[0])
+        bottom_half_sorted = sorted(higher_y_group, key=lambda x: x[0])
+        # Arrange in desired order: upper left, upper right, lower right, lower left
+        centers_sorted = np.float32([top_half_sorted[0], top_half_sorted[-1],
+                                     bottom_half_sorted[-1], bottom_half_sorted[0]])
+        # Transform Perspective
+        M = cv2.getPerspectiveTransform(centers_sorted, pts2) # transformation matrix
+        return M
+
+    def _transform_split(self, img: np.array, m: np.array, filename: str) -> dict[np.array]:
+        """ Takes image and transformation matrix and returns transformed image.
+
+        Args:
+            img (array): image array
+            m (array): transformation matrix
+            filename (str): filename
+
+        Returns:
+            cropped_images (array): transformed image splitted into subsections
+        """
+        transformed_image = cv2.warpPerspective(img, m,
+                                                ((190+ 2* self.offset_mm)*self.mm_to_pixel,
+                                                 (100+ 2* self.offset_mm)*self.mm_to_pixel))
+        # if folder doesn't exist, create it
+        if not os.path.exists(self.path + "/transformed"):
+            os.makedirs(self.path + "/transformed")
+        # Save the image with detected ellipses
+        cv2.imwrite(self.path + f"/transformed/{filename.split(".")[0]}.jpg", transformed_image)
+        # Crop the image
+        cropped_images = {}
+        for i, c in enumerate(self.press_position):
+            # set zero in case it gives a negative number
+            bottom_right_y = (c[1] + 2*self.offset_mm) * self.mm_to_pixel
+            bottom_right_x = (c[0] + 2*self.offset_mm) * self.mm_to_pixel
+            top_left_y = max(bottom_right_y - 2*self.offset_mm*self.mm_to_pixel, 0)
+            top_left_x = max(bottom_right_x - 2*self.offset_mm*self.mm_to_pixel, 0)
```

Comment:
Why do you use 2x offset?

Reply:
True that does not make sense. I changed the offset itself from 20 to 40 and deleted the multiply by 2.